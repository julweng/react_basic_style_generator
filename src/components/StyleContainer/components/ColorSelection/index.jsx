import React, { useState } from "react"
import { func, string, shape } from "prop-types"
import { CompactPicker } from "react-color"
import "./style.less"

const ColorSelection = ({ side, close, color, updateColor }) => {
  const [background, setBackground] = useState(color.backgroundColor)
  const [textColor, setTextColor] = useState(color.color)

  const handleBackgroundChange = color => setBackground(color)
  const handleTextColorChange = color => setTextColor(color)

  const handleSubmitColor = () => {
    const text = typeof textColor === "string" ? textColor : textColor.hex
    const back = typeof background === "string" ? background : background.hex
    updateColor(side, text, back)
    close()
  }

  return (
    <div className={`color__selection__container ${side}`}>
      <div className="color__selection">
        <div className="background__selection">
          <div>Background</div>
          <CompactPicker color={background} onChange={handleBackgroundChange} />
        </div>
        <div className="text__selection">
          <div>Text</div>
          <CompactPicker color={textColor} onChange={handleTextColorChange} />
        </div>
      </div>
      <div className="submit__color__btn">
        <button onClick={close}>Cancel</button>
        <button onClick={handleSubmitColor}>Submit</button>
      </div>
    </div>
  )
}

ColorSelection.propTypes = {
  close: func,
  updateColor: func,
  color: shape({
    color: string,
    backgroundColor: string
  }),
  side: string
}

export default ColorSelection
