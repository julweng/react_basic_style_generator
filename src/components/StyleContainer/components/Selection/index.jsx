import React, { useState } from "react"
import { string, shape, arrayOf, func } from "prop-types"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { faFont, faPalette } from "@fortawesome/free-solid-svg-icons"
import { FontSelection } from "../index"
import "./style.less"

const Selection = ({ type, font, fonts, updateFont, side }) => {
  const [fontSelectionOpen, setFontSelectionOpen] = useState(false)
  //const [openPaletteSelection, setPaletteSelectionOpen] = useState(false)

  const handleFontClose = () => setFontSelectionOpen(false)

  //const handlePaletteClose = () => setPaletteSelectionOpen(false)

  return (
    <div className="selection">
      {type === "font" && (
        <>
          <button
            className="font"
            title="font style"
            onClick={() => setFontSelectionOpen(!fontSelectionOpen)}
          >
            <FontAwesomeIcon icon={faFont} />
          </button>
          {fontSelectionOpen && (
            <FontSelection
              side={side}
              close={handleFontClose}
              font={font}
              fonts={fonts}
              updateFont={updateFont}
            />
          )}
        </>
      )}
      {type === "palette" && (
        <button
          className="palette"
          title="text/background color"
          onClick={() => console.log("hi")}
        >
          <FontAwesomeIcon icon={faPalette} />
        </button>
      )}
    </div>
  )
}

Selection.propTypes = {
  type: string,
  font: shape({}),
  fonts: arrayOf(shape({})),
  updateFont: func,
  side: string
}

export default Selection
