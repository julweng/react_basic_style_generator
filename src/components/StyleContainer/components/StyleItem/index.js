import React from "react"
import { string, shape } from "prop-types"

const StyleItem = ({ side }) => {
  return <div className={side}>Happy picking!</div>
}

StyleItem.propTypes = {
  side: string,
  style: shape({})
}

export default StyleItem
