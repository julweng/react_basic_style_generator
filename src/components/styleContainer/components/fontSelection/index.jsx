import React, { useState } from "react"
import { func, arrayOf, shape, string } from "prop-types"
import { compareVariants, formatRandomFont } from "../../functions"
import { Overlay } from "../index"
import "./style.less"

const FontSelection = ({ close, font, fonts, updateFont, side }) => {
  const [fontOptionsOpen, setFontOptionsOpen] = useState(false)

  const [selectedFont, setSelectedFont] = useState(font)

  const [selectedVariant, setSelectedVariant] = useState(
    Object.keys(font.variants[0])[0]
  )

  const toggle = () => setFontOptionsOpen(!fontOptionsOpen)

  const onChangeFont = (family) => {
    const newFont = formatRandomFont(fonts.filter((f) => f.family === family))
    setSelectedFont(newFont)
    setFontOptionsOpen(false)
  }

  const onChangeVariant = (variant) => {
    setSelectedVariant(variant)
  }

  const handleSubmit = () => {
    updateFont(fonts, side, selectedFont.family, selectedVariant)
    close()
  }

  return (
    <div className={`font__selection__container ${side}`}>
      {fontOptionsOpen && <Overlay />}
      <div className="selected__font" onClick={toggle}>
        <span>{selectedFont.family}</span>
        <span>&#x25BC;</span>
      </div>

      {fontOptionsOpen && (
        <div className="font__options">
          {fonts.map(({ family }) => (
            <li
              key={family}
              value={family}
              onClick={() => onChangeFont(family)}
            >
              {family}
            </li>
          ))}
        </div>
      )}
      <div className="variant__options">
        {selectedFont.variants.map((v) => {
          const [variantKey] = Object.keys(v)
          const [variantVal] = Object.values(v)

          return (
            <div className="font__variant" key={variantKey}>
              <input
                type="radio"
                id={variantKey}
                name={variantKey}
                value={variantKey}
                checked={compareVariants(v, selectedVariant)}
                onChange={(e) => onChangeVariant(e.target.value)}
              />
              <label htmlFor={variantKey}>{variantVal}</label>
            </div>
          )
        })}
      </div>

      <div className="submit__font__btn">
        <button onClick={close}>Cancel</button>
        <button onClick={() => handleSubmit()}>Submit</button>
      </div>
    </div>
  )
}

FontSelection.propTypes = {
  close: func,
  side: string,
  updateFont: func,
  font: shape({
    family: string,
    variants: arrayOf(shape({})),
    variant: shape({})
  }),
  fonts: arrayOf(
    shape({
      family: string,
      variants: arrayOf(string)
    })
  )
}

export default FontSelection
