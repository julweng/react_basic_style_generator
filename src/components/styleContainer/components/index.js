export StyleItem from "./styleItem"
export Selection from "./selection"
export Mask from "./mask"
export FontSelection from "./fontSelection"
export Overlay from "./overlay"
export ColorSelection from "./colorSelection"
export StyleRule from "./styleRule"
