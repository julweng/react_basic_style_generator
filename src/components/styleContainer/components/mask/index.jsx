import { bool } from "prop-types"
import React from "react"
import "./style.less"

const Mask = ({ loaded }) => {
  return <span className={`mask ${loaded ? "loaded" : ""}`} />
}

Mask.propTypes = {
  loaded: bool
}

export default Mask
