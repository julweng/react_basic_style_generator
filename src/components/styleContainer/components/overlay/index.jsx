import React from "react"
import "./style.less"

const Overlay = () => {
  return <div className="overlay" />
}

export default Overlay
