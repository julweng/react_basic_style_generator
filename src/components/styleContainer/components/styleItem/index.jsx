import React from "react"
import { string, shape, arrayOf, func } from "prop-types"
import { Selection, StyleRule } from "../index"
import "./style.less"

const StyleItem = ({ side, style, fonts, font, updateFont, updateColor }) => {
  const defaultText = "Happy Picking"

  return (
    <div className={side} style={style}>
      <div className="selection__container">
        <Selection
          type="font"
          side={side}
          fonts={fonts}
          font={font}
          updateFont={updateFont}
        />
        <Selection
          type="palette"
          side={side}
          color={style}
          updateColor={updateColor}
        />
      </div>
      <p>{font.family || defaultText}</p>
      <StyleRule style={style} />
    </div>
  )
}

StyleItem.propTypes = {
  side: string,
  style: shape({}),
  color: shape({}),
  font: shape({}),
  fonts: arrayOf(shape({})),
  updateFont: func,
  updateColor: func
}

export default StyleItem
