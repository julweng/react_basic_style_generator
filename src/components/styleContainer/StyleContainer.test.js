import React from "react"
import { render, cleanup } from "@testing-library/react"
import "jest-dom/extend-expect"
import { getMockPropsOf } from "__testing__"
import { StyleItem, Mask } from "./components"
import { UnconnectedStyleContainer as StyleContainer } from "./index"

jest.mock("./components/StyleItem", () =>
  jest.fn(() => <div className="StyleItemMock">StyleItemMock</div>)
)

jest.mock("./components/Mask", () =>
  jest.fn(({ loaded }) => (
    <div>{!loaded && <div className="MaskMock">MaskMock</div>}</div>
  ))
)

const createProps = ({ isLoadingFonts = false } = {}) => ({
  leftColor: {
    fontColor: "#000",
    backgroundColor: "#FFF"
  },
  rightColor: {
    fontColor: "#FFF",
    backgroundColor: "#000"
  },
  fonts: [
    {
      family: "Arial",
      category: "sans-serif",
      variants: ["regular", "100italic"]
    },
    {
      family: "Times New Roman",
      category: "serif",
      variants: ["regular", "italic"]
    }
  ],
  leftFont: {
    family: "Arial",
    category: "sans-serif",
    variants: [{ regular: "regular" }, { "100italic": "thin italic" }],
    variant: { regular: "regular" }
  },
  rightFont: {
    family: "Times New Roman",
    category: "serif",
    variants: [{ regular: "regular" }, { italic: "italic" }],
    variant: { regular: "regular" }
  },
  isLoadingFonts,
  updateFont: jest.fn(),
  updateColor: jest.fn(),
  getFonts: jest.fn(),
  getColor: jest.fn()
})

describe("StyleContainer", () => {
  afterEach(cleanup)

  it("Should render without crashing", () => {
    const props = createProps()
    const { container } = render(<StyleContainer {...props} />)
    expect(container).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it("Should render mask when fonts are loading", () => {
    const props = createProps({ isLoadingFonts: true })
    const { container } = render(<StyleContainer {...props} />)
    const mockMask = container.querySelector(".MaskMock")
    expect(mockMask).toBeTruthy()
    const { loaded } = getMockPropsOf(Mask).firstInstance()
    expect(loaded).toBeFalsy()
  })

  it("Should not render mask when fonts finished loading", () => {
    const props = createProps()
    const { container } = render(<StyleContainer {...props} />)
    const mockMask = container.querySelector(".MaskMock")
    expect(mockMask).toBeNull()
  })

  it("Should render header", () => {
    const props = createProps()
    const { container } = render(<StyleContainer {...props} />)
    const header = container.querySelector("h1")
    expect(header).toHaveTextContent("ColoText")
  })

  it("Should render two StyleItem instances", () => {
    const props = createProps()
    const { container } = render(<StyleContainer {...props} />)
    const mockStyleItems = container.querySelectorAll(".StyleItemMock")
    expect(mockStyleItems.length).toEqual(2)
  })

  it("Should render left and right StyleItem", () => {
    const props = createProps()
    render(<StyleContainer {...props} />)
    const left = getMockPropsOf(StyleItem).firstInstance()
    expect(left).toEqual({
      side: "left",
      font: props.leftFont,
      fonts: props.fonts,
      updateFont: props.updateFont,
      updateColor: props.updateColor,
      style: {
        color: "#000",
        backgroundColor: "#FFF",
        fontFamily: "Arial, sans-serif",
        fontWeight: 400,
        fontStyle: "normal"
      }
    })

    const right = getMockPropsOf(StyleItem).lastInstance()
    expect(right).toEqual({
      side: "right",
      font: props.rightFont,
      fonts: props.fonts,
      updateFont: props.updateFont,
      updateColor: props.updateColor,
      style: {
        color: "#FFF",
        backgroundColor: "#000",
        fontFamily: "Times New Roman, serif",
        fontWeight: 400,
        fontStyle: "normal"
      }
    })
  })

  it("Should render footer", () => {
    const props = createProps()
    const { container } = render(<StyleContainer {...props} />)
    const footer = container.querySelector("p")
    expect(footer).toHaveTextContent("2019 Made with React by Jul W")
  })
})
