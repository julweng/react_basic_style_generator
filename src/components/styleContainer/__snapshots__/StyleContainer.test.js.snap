// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleContainer Should render without crashing 1`] = `
<div>
  <div
    class="main"
  >
    <div />
    <header>
      <h1>
        ColoText
      </h1>
    </header>
    <div
      class="style__container"
    >
      <div
        class="StyleItemMock"
      >
        StyleItemMock
      </div>
      <div
        class="StyleItemMock"
      >
        StyleItemMock
      </div>
    </div>
    <footer>
      <p>
        Â© 2019 Made with React by Jul W
      </p>
    </footer>
  </div>
</div>
`;
